framework/src/convergence/ResidualConvergence.C

```diff
@@ -0,0 +1,535 @@
+  }
+  else
+    fnorm_old = system._last_nl_rnorm;
+
+  // Check for nonlinear residual pingpong.
+  // Pingpong will always start from a residual increase
+  if ((_n_nl_pingpong % 2 == 1 && !(fnorm_petsc > fnorm_old)) ||
+      (_n_nl_pingpong % 2 == 0 && fnorm_petsc > fnorm_old))
+    _n_nl_pingpong += 1;
+  else
+    _n_nl_pingpong = 0;
+
+  long int _nl_forced_its = _fe_problem.getNonlinearForcedIterations();
+
+  std::ostringstream oss;
+  if (fnorm_petsc != fnorm_petsc)
+  {
+    oss << "Failed to converge, function norm is NaN\n";
+    reason = MooseConvergenceStatus::DIVERGED;
+  }
+  else if ((it_petsc >= _nl_forced_its) && fnorm_petsc < _atol)
+  {
+    oss << "Converged due to function norm " << fnorm_petsc << " < " << _atol << '\n';
+    reason = MooseConvergenceStatus::CONVERGED;
+  }
+  else if (_nfuncs >= _maxf)
+  {
+    oss << "Exceeded maximum number of function evaluations: " << _nfuncs << " > " << _maxf << '\n';
+    reason = MooseConvergenceStatus::DIVERGED;
+  }
+  else if ((it_petsc >= _nl_forced_its) && it_petsc && fnorm_petsc > system._last_nl_rnorm &&
+           fnorm_petsc >= _div_threshold)
+  {
+    oss << "Nonlinear solve was blowing up!\n";
+    reason = MooseConvergenceStatus::DIVERGED;
+  }
+  if ((it_petsc >= _nl_forced_its) && it_petsc && reason == MooseConvergenceStatus::ITERATING)
+  {
+    // Set the reference residual depending on what the user asks us to use.
+    const auto the_residual = system.referenceResidual();
+    if (checkRelativeConvergence(it_petsc, fnorm_petsc, the_residual, _rtol, _atol, oss))
+      reason = MooseConvergenceStatus::CONVERGED;
+    else if (snorm_petsc < _stol * xnorm_petsc)
+    {
+      oss << "Converged due to small update length: " << snorm_petsc << " < " << _stol << " * "
+          << xnorm_petsc << '\n';
+      reason = MooseConvergenceStatus::CONVERGED;
+    }
+    else if (_divtol > 0 && fnorm_petsc > the_residual * _divtol)
+    {
+      oss << "Diverged due to initial residual " << the_residual << " > divergence tolerance "
+          << _divtol << " * initial residual " << the_residual << '\n';
+      reason = MooseConvergenceStatus::DIVERGED;
+    }
+    else if (_nl_abs_div_tol > 0 && fnorm_petsc > _nl_abs_div_tol)
+    {
+      oss << "Diverged due to residual " << fnorm_petsc << " > absolute divergence tolerance "
+          << _nl_abs_div_tol << '\n';
+      reason = MooseConvergenceStatus::DIVERGED;
+    }
+    else if (_n_nl_pingpong > _n_max_nl_pingpong)
+    {
+      oss << "Diverged due to maximum nonlinear residual pingpong achieved" << '\n';
+      reason = MooseConvergenceStatus::DIVERGED;
+    }
+  }
+
+  system._last_nl_rnorm = fnorm_petsc;
+  system._current_nl_its = static_cast<unsigned int>(it_petsc);
+
+  std::string msg;
+  msg = oss.str();
+  if (_app.multiAppLevel() > 0)
+    MooseUtils::indentMessage(_app.name(), msg);
+  if (msg.length() > 0)
+#if !PETSC_VERSION_LESS_THAN(3, 17, 0)
+    PetscInfo(snes, "%s", msg.c_str());
```

Comment:
```suggestion
    LibmeshPetscCall(PetscInfo(snes, "%s", msg.c_str()));
```
with the recent PETSc update, we do not allow discarding the `PetscErrorCode` return of PETSc functions

Reply:
this suggestion should fix the errors from CIVET